### PR TITLE
`Quiz exercises`: Fix batch quiz editing when batch is added

### PR DESCRIPTION
**Pull request by rriyaldhi**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

[#5958](https://github.com/ls1intum/Artemis/issues/5958)

https://user-images.githubusercontent.com/10885503/207876795-1db3c723-75a2-4d6c-9a52-3d09faa6cbe5.mov

### Description

- Disabled edit button if batch is added for batched quiz

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course

1. Log in to Artemis
2. Navigate to Course Administration
3. Select the Course
4. Click Exercises
5. Create Quiz Exercise
6. Set to Batch
7. Click Save
8. Click Add Batch
9. Verify that Edit button is disabled

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| quiz-exercise.component.ts | 63.41% | 88.61% |

### Screenshots

https://user-images.githubusercontent.com/10885503/207878483-ff3a6b94-8bec-45d5-8fc4-64920e020173.mov

### CODE REVIEW

**Review by PhilipWerz — APPROVED** [review]
I created two quiz exercises. For the first one I set it to batched from the beginning on. For the other one I changed it after creating the exercise. In both cases edit button was activated prior to clicking on "add batch" and was **deactivated** after clicking on "add batch". Also tried to click "add batch" multiple times and it still worked. So everything worked as it should. Tested on TS1

**Review by laadvo — DISMISSED** [review]
The edit button is disabled after a batch has been added to the quiz as expected. Tested on ts2

**Review by LiaLism — APPROVED** [review]
Edit button is disabled. Everything worked as expected. Tested on ts1.

File: src/main/webapp/app/exercises/quiz/manage/quiz-exercise.component.ts [hunk 1]
```diff
@@ -314,4 +314,18 @@ export class QuizExerciseComponent extends ExerciseComponent {
...
     }
+
+    /**
+     * Check if quiz is editable
+     * @param quizExercise the quiz exercise which will be checked
+     * @return {boolean} true if the quiz is editable and false otherwise
+     */
+    isQuizEditable(quizExercise: QuizExercise): boolean {
+        if (quizExercise.quizMode === QuizMode.BATCHED) {
+            if (quizExercise.quizBatches?.length ?? 0 > 0) {
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/quiz/manage/quiz-exercise.component.ts:325]: 0 is falsy so this should be the same:
```suggestion
            if (quizExercise.quizBatches?.length) {
```

Comment (rriyaldhi) [reply src/main/webapp/app/exercises/quiz/manage/quiz-exercise.component.ts]: Thanks for the suggestion

**Review by laadvo — DISMISSED** [review]
The edit button works as intended, can reapprove. Tested on ts2

**Review by timo07k — APPROVED** [review]
Works as expected. Tested on TS1.

**Review by JohannesStoehr — COMMENTED** [review]
Calling functions in the html often leads to the problem that these functions are called over and over again hurting the performance. Did you make sure that this is not the case here?
Otherwise you could probably store this in a local variable and initialize it in onInit or something

**Comment by rriyaldhi** [comment]
> Calling functions in the html often leads to the problem that these functions are called over and over again hurting the performance. Did you make sure that this is not the case here? Otherwise you could probably store this in a local variable and initialize it in onInit or something

Hi @JohannesStoehr , thanks for noticing. Yes, the function keeps getting called. I updated the code so that the function is called only when necessary.

File: src/main/webapp/app/exercises/quiz/manage/quiz-exercise.component.ts [hunk 2]
```diff
@@ -314,4 +316,18 @@ export class QuizExerciseComponent extends ExerciseComponent {
...
         );
     }
+
+    /**
+     * Check if quiz is editable
+     * @param quizExercise the quiz exercise which will be checked
+     * @return {boolean} true if the quiz is editable and false otherwise
+     */
+    isQuizEditable(quizExercise: QuizExercise): boolean {
+        if (quizExercise.quizMode === QuizMode.BATCHED) {
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/exercises/quiz/manage/quiz-exercise.component.ts:326]: Could combine the nested if statements

File: src/main/webapp/app/entities/quiz/quiz-exercise.model.ts [hunk 3]
```diff
@@ -51,6 +51,7 @@ export class QuizExercise extends Exercise {
 
     public isActiveQuiz?: boolean;
     public isPracticeModeAvailable?: boolean;
+    public isEditable?: boolean;
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/entities/quiz/quiz-exercise.model.ts:54]: Add default value and comment in constructors as with the other ones maybe?

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session on ts1. Works as expected.

**Review by danielbode — APPROVED** [review]
Tested on ts1 and it works!

**Review by Yunqing-Wang — APPROVED** [review]
Works perfectly, tested in the test session

Yunqing

File: src/main/webapp/app/entities/quiz/quiz-exercise.model.ts (hunk 3 above)
Comment (rriyaldhi) [reply src/main/webapp/app/entities/quiz/quiz-exercise.model.ts:54]: Thanks for the suggestion. I added the default value and comment in the constructor.

File: src/main/webapp/app/exercises/quiz/manage/quiz-exercise.component.ts (hunk 2 above)
Comment (rriyaldhi) [reply src/main/webapp/app/exercises/quiz/manage/quiz-exercise.component.ts]: I combined the if statements.

**Review by laadvo — APPROVED** [review]
The quiz cannot be edited when a batch has been added. Tested on ts1

**Review by LiaLism — APPROVED** [review]
Edit button is disabled. Tested on ts1.

**Review by Santia-go — APPROVED** [review]
Tested on TS0
The edit button is disabled when the batch is added, and when the quiz ends, the mentioned button becomes "re-evaluate". 
Approved!

**Review by pellumbbaboci — APPROVED** [review]
Tested on TS1, works as expected and code changes look good.

**Comment by pellumbbaboci** [comment]
@rriyaldhi i noticed that we can still access the quiz page and edit it. Do we need to handle that as well? See video

https://user-images.githubusercontent.com/26312757/209972696-cc109499-5452-4cd0-81cc-54abb0a24201.mp4

**Comment by rriyaldhi** [comment]
> @rriyaldhi i noticed that we can still access the quiz page and edit it. Do we need to handle that as well? See video
> 
> 
> 
> https://user-images.githubusercontent.com/26312757/209972696-cc109499-5452-4cd0-81cc-54abb0a24201.mp4
> 
> 
> 
> 

Thanks for noticing! Since it happens not because of this PR, I'll create a follow-up PR to fix it.

**Pull request merged**